### PR TITLE
To add Rectified Adam Algorithm to Optimizers

### LINKED ISSUE #24892: Implement RAdam optimizer ?
## 🚀 Feature
implement RAdam optimizer ? 

## Motivation
I've recently come across this paper on rectified Adam, which shows a clear improvement on the existing Adam optimizer. There is a issue on the tensorflow github page for a feature request and I thought pytorch definitely needs someone to bring it up. 

## Pitch
Being able to use RAdam in Pytorch. 

## Additional context

RAdam Paper : https://arxiv.org/pdf/1908.03265v1.pdf

cc @vincentqb

### PR DESCRIPTION
**Pull request by iramazanli**
Fixes : #24892

In the paper : https://arxiv.org/pdf/1908.03265.pdf  Liyuan Liu et al. suggested a new optimization algorithm with an essence of similar to Adam Algorithm.

It has been discussed in the paper that, without warmup heuristic, in the early stage of adaptive optimization / learning algorithms sometimes we can get undesirable large variance which can slow overall convergence process. 

Authors proposed the idea of rectification of variance of adaptive learning rate when it is expected to be high.

Differing from the paper, we selected variance tractability cut-off as 5 instead of 4. This adjustment is common practice, and could be found in the code-repository and also tensorflow swift optim library as well : 

https://github.com/LiyuanLucasLiu/RAdam/blob/2f03dd197022da442c6a15c47321f4335d113a3f/radam/radam.py#L156

https://github.com/tensorflow/swift-apis/blob/f51ee4618d652a2419e998bf9418ad80bda67454/Sources/TensorFlow/Optimizers/MomentumBased.swift#L638

### CODE REVIEW

File: torch/optim/_functional.py [hunk 1]
```diff
@@ -357,3 +357,51 @@ def asgd(params: List[Tensor],
...
+        # Decay the first and second moment running average coefficient
+        exp_avg.mul_(beta1).add_(grad, alpha=1 - beta1)
+        exp_avg_sq.mul_(beta2).addcmul_(grad, grad, value=1 - beta2)
+
+        # maximum length of the approximated SMA
+        rho_inf = 2 / (1 - beta2) - 1
+        # compute the length of the approximated SMA
+        rho_t = rho_inf - 2 * step * (beta2 ** step) / bias_correction2
+
+        if rho_t > 5.:
```
Comment (vincentqb) [inline torch/optim/_functional.py:399]: The value I see in algorithm 2 of the paper say "e.g. 4". This suggests (1) the default value should be 4 and not 5? and (2) the value should be parametrizable.

File: torch/optim/_functional.py [hunk 2]
```diff
@@ -357,3 +357,51 @@ def asgd(params: List[Tensor],
...
+            # Compute the variance rectification term and update parameters accordingly
+            rect = math.sqrt((rho_t - 4) * (rho_t - 2) * rho_inf / ((rho_inf - 4) * (rho_inf - 2) * rho_t))
```
Comment (vincentqb) [inline torch/optim/_functional.py:401]: are this integers (e.g. 4, 2) related in to the threshold for rho_t?

File: torch/optim/_functional.py [hunk 3]
```diff
@@ -357,3 +357,51 @@ def asgd(params: List[Tensor],
...
+            denom = (exp_avg_sq.sqrt() / math.sqrt(bias_correction2))
+            step_size = lr * rect / bias_correction1
+            param.addcdiv_(exp_avg, denom, value=-step_size)
+        else:
+            step_size = lr / bias_correction1
```
Comment (vincentqb) [inline torch/optim/_functional.py:406]: based on [algorithm 2](https://arxiv.org/pdf/1908.03265.pdf), shouldn't `lr` be momentum here?

Comment (vincentqb) [reply torch/optim/_functional.py:406]: the only difference between the two case should be `r_t l_t`. can you write the code in a way that makes this more explicit?

File: torch/optim/_functional.py (hunk 1 above)
Comment (iramazanli) [reply torch/optim/_functional.py:399]: I completely agree that the number is 4 in the paper.. however, authors of the paper implemented it to be 5. in the corresponding repository for the paper giving the reasoning 4  is an estimated number for cut-off and 5 is safer choice:
 https://github.com/LiyuanLucasLiu/RAdam/blob/2f03dd197022da442c6a15c47321f4335d113a3f/radam/radam.py#L156

We can see TensorFlow Swift also followed the same practice : 

https://github.com/tensorflow/swift-apis/blob/f51ee4618d652a2419e998bf9418ad80bda67454/Sources/TensorFlow/Optimizers/MomentumBased.swift#L638

File: torch/optim/_functional.py (hunk 2 above)
Comment (iramazanli) [reply torch/optim/_functional.py:401]: In practice, we can see that they are not super related. As you can see both author of the paper and Swift implementors use the threshold cut-off to be 5 however, 4 and 2 remained the same.

Speaking of theory, authors made analysis based on assumption rho_t >= 4. Basically, it means that relation is the algorithm is valid for cutoff to be larger or equal than 4. Therefore, whatever cut-off we select (4, 5 or etc) these pair (4,2) will still be the same independently. However, we need to keep in mind, that if we select cut-off to be less than 4 the algorithm/ its analysis is no more valid.

File: torch/optim/_functional.py (hunk 1 above)
Comment (vincentqb) [reply torch/optim/_functional.py:399]: Ok, please add these references in the description of the PR to justify the choice of 5.

Comment (iramazanli) [reply torch/optim/_functional.py:402]: Sounds great :) I am adding it now

File: torch/optim/_functional.py (hunk 3 above)
Comment (iramazanli) [reply torch/optim/_functional.py]: Yes, i agree that there is a confusion with the variable names here and showing the difference is only r_t l_t would be much better presentation.

Therefore I adjusted the code accordingly. (my apologies for previous confusion, actually lr here is step-size in the paper, but now code should be much more readable)

**Review by vincentqb — APPROVED** [review]
LGTM!

**Comment by vincentqb** [comment]
discussed offline: we'll keep the name as RAdam instead of [PlainRAdam](https://github.com/LiyuanLucasLiu/RAdam/blob/master/radam/radam.py#L96). If we need another version, we could explore calling it something else (ModifiedRAdam for [this one](https://github.com/LiyuanLucasLiu/RAdam/blob/master/radam/radam.py#L5)?) or have a toggle?

**Comment by samestep** [comment]
Reverting because this broke doc build:

- https://github.com/pytorch/pytorch/runs/2887654727
- https://app.circleci.com/pipelines/github/pytorch/pytorch/340103/workflows/ae9cd45b-39ff-4f74-9eb4-9528bd5cbf5f/jobs/14302190
- https://app.circleci.com/pipelines/github/pytorch/pytorch/340103/workflows/ae9cd45b-39ff-4f74-9eb4-9528bd5cbf5f/jobs/14302198
